elastic/apm-server
# display timeouts to user as "server timeout"

## Motivation/summary

timeouts configured at the server level result in
the request context being canceled; translate this
to a "server timeout" message before
logging/sending to the client.

~~the actual message being sent and status code
could also be changed, if there's a better way to
communicate what's happening, or a more
appropriate status code, we can change it. i'm not
sure at which point in the request/response
lifecycle it's failing, so it could be a 408 or
503.~~

~~note: this feels like a very broad approach.
reading around online did not provide a concrete
way to check that the request had timed out. an
alternative method, such as setting a
context.WithTimeout on the request instead of the
server's ReadTimeout might give us more visibility
into the cause. my concern with the approach in
this PR is that some un-related error might cause
the request (and context) to be canceled, and then
we report this as a timeout.~~

EDIT:

the translation is now implemented as a middleware that is applied to all endpoints.

the clients receive a 503 service unavailable, with text "request timed out".

fixes https://github.com/elastic/apm-server/issues/4484

## Checklist




## How to test these changes

build this branch locally and run with `./apm-server -c apm-server.secured.yml -e 2>&1 | grep 'timed out'`

a user can build the code in this branch, and run `hey-apm -instances 5`, or whatever is an appropriate `-instances` count to overload the server, causing requests to timeout.

expected log lines should look something along the lines of:

```
{
  "log.level": "error",
  "@timestamp": "2021-03-05T10:56:57.775+0100",
  "log.logger": "request",
  "log.origin": {
    "file.name": "middleware/log_middleware.go",
    "file.line": 60
  },
  "message": "request timed out",
  "url.original": "/intake/v2/events?verbose=",
  "http.request.method": "POST",
  "user_agent.original": "hey-apm",
  "source.address": "127.0.0.1",
  "http.request.id": "f4fbb971-4d88-494f-90b9-bbc31ad3f7df",
  "event.duration": 30561586246,
  "http.response.status_code": 503,
  "error.message": "request timed out",
  "ecs.version": "1.6.0"
}
```
